gateway/src/node/client_handling/websocket/connection_handler/ecash/credential_sender.rs

```diff
@@ -0,0 +1,216 @@
+            //no proposal for freepasses
+            return Ok(());
+        }
+        let res = self
+            .nyxd_client
+            .write()
+            .await
+            .prepare_credential(
+                pending.credential.data.serial_number_b58(),
+                pending.address.to_string(),
+                None,
+            )
+            .await?;
+        let proposal_id = find_attribute(&res.logs, "wasm", BANDWIDTH_PROPOSAL_ID)
+            .ok_or(RequestHandlingError::ProposalIdError {
+                reason: String::from("proposal id not found"),
+            })?
+            .value
+            .parse::<u64>()
+            .map_err(|_| RequestHandlingError::ProposalIdError {
+                reason: String::from("proposal id could not be parsed to u64"),
+            })?;
+
+        let proposal = self
+            .nyxd_client
+            .read()
+            .await
+            .query_proposal(proposal_id)
+            .await?;
+        if !pending
+            .credential
+            .matches_serial_number(&proposal.description)?
+        {
+            return Err(RequestHandlingError::ProposalIdError {
+                reason: String::from("proposal has different serial number"),
+            });
+        }
+        pending.proposal_id = Some(proposal_id);
+        Ok(())
+    }
+
+    async fn send_credential(pending: &mut PendingCredential) -> Result<(), RequestHandlingError> {
+        if pending.credential.data.typ.is_ticketbook() && pending.proposal_id.is_none() {
+            return Err(RequestHandlingError::ProposalIdError {
+                reason: "proposal id is absent".to_string(),
+            });
+        }
+        let request = VerifyEcashCredentialBody::new(
+            pending.credential.data.clone(),
+            pending.address.clone(),
+            pending.proposal_id,
+        );
+        let mut failed_api = Vec::new();
+        for client in &pending.api_clients {
+            match client.verify_offline_credential(&request).await {
+                Ok(response) => {
+                    //Even if credential isn't accepted, we did contact the validator and resubmitting the same credential won't change anything. We can consider the sending as done
+                    if response != VerifyEcashCredentialResponse::Accepted {
+                        log::debug!(
+                            "Validator {} didn't accept the credential - Reason : {}",
+                            client.nym_api.current_url(),
+                            response
+                        );
+                    }
+                }
+                Err(e) => {
+                    log::warn!("Validator {} could not be reached. There might be a problem with the coconut endpoint - {:?}", client.nym_api.current_url(), e);
+                    failed_api.push(client.clone());
+                }
+            }
+        }
+        pending.api_clients = failed_api;
+        if pending.api_clients.is_empty() {
```

Comment:
do we have to have **ALL** or is threshold sufficient? we should have some resilience, no?

Reply:
`threshold` should be sufficient, but the more the better I guess